var/spack/repos/builtin/packages/py-tensorflow-probability/package.py

```diff
@@ -48,9 +58,27 @@ class PyTensorflowProbability(Package):
     depends_on("py-dm-tree", when="@0.12:", type=("build", "run"))
 
     # tensorflow_probability/python/__init__.py
-    depends_on("py-tensorflow@2.10:", when="@0.18:", type=("build", "run"))
-    depends_on("py-tensorflow@2.4:", when="@0.12:", type=("build", "run"))
-    depends_on("py-tensorflow@1.14:", when="@0.8:", type=("build", "run"))
+    # TODO: reactivate the JAX versions once the JAX package is available with newer versions
+    #  also add jaxlib as a dependency
+    # TODO: reactivate once TF 2.15 is ready https://github.com/spack/spack/pull/41069
+    # depends_on("py-tensorflow@2.15", when="@0.23", type=("build", "run"))
+    # depends_on("py-jax@0.4.20:0.4", when="@0.23", type=("build", "run"))
+
+    depends_on("py-tensorflow@2.14", when="@0.22", type=("build", "run"))
```

Comment:
I believe these should be >=, not ==. At least, that's the way things are encoded in `__init__.py`

Reply:
They do, true, but in the [changelog](https://github.com/tensorflow/probability/releases/tag/v0.22.1) they say against which version things are tested and stable. The problem with keeping only a lower limit is that this may break versions with future TF releases as painfully happenend multiple times in the past. Therefore, we adopted in conda-forge a strict policy towards TF (and a laxer one towards JAX [after some discussion](https://github.com/conda-forge/tensorflow-probability-feedstock/issues/45)).

If you have a strong opinion on it, please go ahead and adjust (or let me know), both have advantages and disadvantages, so fine with both, but if we choose to relax, the question is what to do if users run into it? Add upper boundaries? If so where? Should we test all versions of TFP against TF, something that not even TFP seems to be doing?

Reply:
I don't have strong opinions here. I've also needed to pin some things related to TF/JAX in other packages because they frequently break backwards compatibility. One of the main advantages of Spack over pip/conda is that it's trivially easy to add upper bounds to ancient versions if users discover incompatibilities. So some of the conda-forge discussion may not exactly apply here. I'll leave the decision up to you.